logic/src/commonMain/kotlin/com/wire/kalium/logic/functional/Flow.kt

```diff
@@ -0,0 +1,21 @@
+package com.wire.kalium.logic.functional
+
+import kotlinx.coroutines.flow.Flow
+import kotlinx.coroutines.flow.flow
+import kotlinx.coroutines.flow.take
+
+suspend fun <A, B> Collection<A>.flatMapFromIterable(
+    block: suspend (A) -> Flow<B>
+): Flow<List<B>> {
+    return flow {
+        val result = mutableListOf<B>()
+        this@flatMapFromIterable.forEach { a ->
+            block(a)
+                .take(1)
+                .collect { b ->
+                    result.add(b)
+                    if (result.size == this@flatMapFromIterable.size) emit(result)
+                }
+        }
+    }
+}
```

Comment:
Niceee method. Small question, doesn't `take(1)` do the same than `first()` ?

Reply:
good question :) 
from what i got: `take(1)` returns `Flow<T>`, so it will take 1 emit of that flow (when it comes), pass it to the `collect{}` and abort the Flow. 
On the other hand:`first()` returns not `Flow<T>` but `T`, so it kinda waiting for the original Flow to emit. 

So in current example: if I need to fetch list of 1000 elements: 
- `take(1)` won't wait for each element to be fetched, but request it, and **when** it comes - add it to the `result` List
- `first()` will wait for each fetching to add it to the `result` List and only after that request for the next element 

If I'm right (which I'm not 100% sure, as have not so deep knowledge in Coroutines) `take(1)` will work faster OR with the same speed as `first()`.

Reply:
The main difference are:

### Take
- Is not an terminal operator. It returns another flow.
- Once you call it, nothing happens. You still need to call something like `collect` or another terminal operator in order to start the flow.
- Is silent, if the flow is closed without emitting any value, the same behaviour will be propagated to whoever is calling take, without exceptions.

### First
- Is a terminal operator. It doesn't return another flow.
- Once you call it, the flow will start immediatelly and hang until the value is emmited.
- If the flow is closed without emitting a value, it will throw `NoSuchElementException`.


---- 
They are similar in the sense that they will make the caller wait for a value to be emitted when collecting the flow. Until the value is emitted, it hangs.

Reply:
In this case, I don't understand how can this code work 🧐

`take(1)` will make sure that only ONE value is collected. If you run this with two or more items, the flow will never emmit enything, no?

Reply:
All things considered, because we have a `take(1)` quickly followed by a `collect()`, it doesn't make much of a difference between using it `take` or `first`.

We can re-write the function to be like this:

```kt
suspend inline fun <A, B> Collection<A>.flatMapFromIterable(
    crossinline block: suspend (A) -> Flow<B>
): Flow<List<B>> = flow {
    val result = mutableListOf<B>()

    if (isEmpty()) emit(result)

    forEach { a ->
        result.add(block(a).first())
    }
    emit(result)
}
```